protoc-gen-php: Support for package namespaces

cypres opened the pull request:
Been working on rudimentary support for package based namespaces.

It needs some additional work before it should/could be merged, but I thought I would post it anyway. Especially the following would be nice:
- [PSR-0](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-0.md) compliance
- Support for nested types with namespaces instead of underscores (almost requires PSR-0)
- A way to disable the feature via (php) option or otherwise

It works though, so if you really need/want namespaces it's usable.

cypres committed:
Support for package namespaces
Replaced the (php).namespace feature with namespaces from the package.
In addition the (php).namespace feature didn't work as intended since
base classes was not properly referenced etc.
There are some major caveats
 - The compiler is not PSR-0 compliant
 - Nested types are still supported, but with underscores.
 - Currently no way to disable the feature

chobie commented:
Hi, this is nice PR! 

> - A way to disable the feature via (php) option or otherwise

do you have some time to implement this feature?

cypres commented:
Sure I can add way to toggle it on and off. I need to work out how to keep both features with minimal code impact, but I'll work on it next week. Preferably while also keeping support for (php).namespace option.

I should also add some new tests if you want to merge it.

cypres committed:
Re-worked namespaces, optional feature and fixes
Added optional feature to generate packaged based namespaces, defaults
to false to ensure compatibility.
Reverted multiple changes to apply more cleanly to master branch.
Fixed extensions

cypres commented:
Sorry for the delay. I'm pretty tied up currently. I've re-worked the feature to apply more cleanly, and for it to be optional. 
Unit testing still missing atm.

cypres committed:
Merged upstream changes

chobie commented:
@duedal Thank you. It looks good to me. 
I want testcases. But It does not required atm as protoc-gen-php don't have it.

could you send another unit test PR when you have time?

cypres commented:
Thank you, will send a PR with tests later, if all goes well I should be able to find the time next week.

chobie commented:
@duedal btw, what testing framework do you use for this?
I'd like to add gtest for conformance test. Please tell me current status
